src/main/offline/io/stargate/sgv2/jsonapi/service/cqldriver/sstablewriter/OfflineCommandsProcessor.java

```diff
@@ -0,0 +1,203 @@
+package io.stargate.sgv2.jsonapi.service.cqldriver.sstablewriter;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.micrometer.core.instrument.simple.SimpleMeterRegistry;
+import io.smallrye.config.SmallRyeConfig;
+import io.smallrye.config.SmallRyeConfigBuilder;
+import io.stargate.sgv2.api.common.config.MetricsConfig;
+import io.stargate.sgv2.jsonapi.api.model.command.CommandContext;
+import io.stargate.sgv2.jsonapi.api.model.command.CommandResult;
+import io.stargate.sgv2.jsonapi.api.model.command.impl.*;
+import io.stargate.sgv2.jsonapi.api.request.DataApiRequestInfo;
+import io.stargate.sgv2.jsonapi.api.response.BeginOfflineSessionResponse;
+import io.stargate.sgv2.jsonapi.api.response.EndOfflineSessionResponse;
+import io.stargate.sgv2.jsonapi.api.response.OfflineGetStatusResponse;
+import io.stargate.sgv2.jsonapi.api.response.OfflineInsertManyResponse;
+import io.stargate.sgv2.jsonapi.config.DocumentLimitsConfig;
+import io.stargate.sgv2.jsonapi.config.OperationsConfig;
+import io.stargate.sgv2.jsonapi.service.cqldriver.CQLSessionCache;
+import io.stargate.sgv2.jsonapi.service.cqldriver.executor.QueryExecutor;
+import io.stargate.sgv2.jsonapi.service.embedding.DataVectorizerService;
+import io.stargate.sgv2.jsonapi.service.embedding.operation.EmbeddingProvider;
+import io.stargate.sgv2.jsonapi.service.processor.CommandProcessor;
+import io.stargate.sgv2.jsonapi.service.resolver.CommandResolverService;
+import io.stargate.sgv2.jsonapi.service.resolver.model.impl.BeginOfflineSessionCommandResolver;
+import io.stargate.sgv2.jsonapi.service.resolver.model.impl.EndOfflineSessionCommandResolver;
+import io.stargate.sgv2.jsonapi.service.resolver.model.impl.OfflineGetStatusCommandResolver;
+import io.stargate.sgv2.jsonapi.service.resolver.model.impl.OfflineInsertManyCommandResolver;
+import io.stargate.sgv2.jsonapi.service.shredding.Shredder;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class OfflineCommandsProcessor {
+  private static final Logger logger = LoggerFactory.getLogger(OfflineCommandsProcessor.class);
+  private OperationsConfig operationsConfig;
+  private CQLSessionCache cqlSessionCache;
+  private CommandResolverService commandResolverService;
+  private DataVectorizerService dataVectorizerService;
+  private static OfflineCommandsProcessor offlineCommandsProcessor;
+  private static boolean initialized;
+
+  private OfflineCommandsProcessor() {}
+
+  public static OfflineCommandsProcessor getInstance() {
```

Comment:
Instead of separate `init()`, could just have

```
   private final OfflineCommandsProcessor offlineCommandsProcessor = new OfflineCommandsProcessor();

   public static OfflineCommandsProcessor getInstance() { return offlineCommandsProcessor; }
```

and do everything `init()` has now in constructor? That would change most fields to be `final` as well.

Reply:
I still have the `init` method for handling the synchronization. I'll test and remove that if classloader initialization takes care of it (I added a TODO marker there for now).